File: backend/app/services/participant_service.py
from uuid import UUID

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.infrastructure.models import ParticipantModel
from app.repositories import participant_repository
from app.schemas.participant_schema import ParticipantCreate, ParticipantUpdate


def list_participants(db: Session) -> list[ParticipantModel]:
    return participant_repository.get_all(db)


def get_participant(db: Session, participant_id: UUID) -> ParticipantModel:
    participant = participant_repository.get_by_id(db, participant_id)
    if participant is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Participant {participant_id} not found",
        )
    return participant


def create_participant(db: Session, data: ParticipantCreate) -> ParticipantModel:
    return participant_repository.create(db, data)


def update_participant(
    db: Session, participant_id: UUID, data: ParticipantUpdate
) -> ParticipantModel:
    participant = participant_repository.update(db, participant_id, data)
    if participant is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Participant {participant_id} not found",
        )
    return participant


def delete_participant(db: Session, participant_id: UUID) -> None:
    deleted = participant_repository.delete(db, participant_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Participant {participant_id} not found",
        )

File: frontend/src/services/participantService.ts
import { request } from "../api/client";
import type {
  ParticipantCreate,
  ParticipantRead,
  ParticipantUpdate,
} from "../types/api";

export function getParticipants(): Promise<ParticipantRead[]> {
  return request<ParticipantRead[]>("/participants");
}

export function getParticipant(id: string): Promise<ParticipantRead> {
  return request<ParticipantRead>(`/participants/${id}`);
}

export function createParticipant(
  data: ParticipantCreate,
): Promise<ParticipantRead> {
  return request<ParticipantRead>("/participants", { method: "POST", body: data });
}

export function updateParticipant(
  id: string,
  data: ParticipantUpdate,
): Promise<ParticipantRead> {
  return request<ParticipantRead>(`/participants/${id}`, {
    method: "PUT",
    body: data,
  });
}

export function deleteParticipant(id: string): Promise<void> {
  return request<void>(`/participants/${id}`, { method: "DELETE" });
}
